test/input_gen/recorder.py

```diff
@@ -0,0 +1,127 @@
+        loss_fn=None,
+        optimizer=tf.keras.optimizers.SGD(lr=1.0),
+    ):
+        self.inputs = inputs
+        self.outputs = outputs
+        self.model = K.Model(inputs=inputs, outputs=outputs)
+        self.loss_fn = loss_fn
+        self.optimizer = optimizer
+        self.file = open(file_name, "wb")
+        self.generate_data(input_shape, label_shape)
+
+    def __del__(self):
+        self.file.close()
+
+    def _rand_like(self, tensorOrShape, scale=10):
+        try:
+            return tf.random.uniform(tensorOrShape.shape, dtype=tf.float32) * scale
+        except AttributeError:
+            return tf.random.uniform(tensorOrShape, dtype=tf.float32) * scale
+
+    ##
+    # @brief generate data using uniform data from a function and save to the file.
+    # @note one-hot label is supported for now, this could be extended if needed.
+    def generate_data(self, input_shape, label_shape):
+        """This part loads data, should be changed if you are gonna load real data"""
+        self.initial_input = self._rand_like(input_shape)
+        self.label = tf.one_hot(indices=[1] * label_shape[0], depth=label_shape[1])
+
+        self.initial_input.numpy().tofile(self.file)
+        self.label.numpy().tofile(self.file)
+
+    def _write_items(self, *items):
+        for item in items:
+            print(item)
+            try:
+                item.numpy().tofile(self.file)
+            except AttributeError:
+                pass
+            try:
+                print(item.shape, " data is generated")
+            except:
+                pass
+
+    ##
+    # @brief model iteration wrapper that listen to the gradient and outputs of the model
+    # each results are recorded.
+    def step(self):
+        self.model.summary()
+
+        with tf.GradientTape(persistent=True) as tape:
+            tape.watch(self.initial_input)
+            outputs = self.model(self.initial_input)
+
+            if self.loss_fn:
+                loss = self.loss_fn(self.label, outputs[-1])
+                outputs.append(loss)
+                print("loss is %s" % loss)
+
+        results = [self.initial_input] + outputs
+
+        for idx, layer in enumerate(self.model.layers):
+            print("generating for %s" % layer.name)
+
+            weights = layer.trainable_weights.copy()
+            gradients = tape.gradient(results[-1], layer.trainable_weights)
+            dweights = tape.gradient(results[-1], results[idx])
+
+            # input, weights, gradients, output, dx
+            # you should take weight order to account (eg. I think conv2d has different order)
+            self._write_items(
+                *[results[idx], *weights, *gradients, results[idx + 1], dweights]
+            )
+
+            self.optimizer.apply_gradients(zip(gradients, layer.trainable_weights))
+
+        self._write_items(results[-1])
+
+    ##
+    # @brief run function
+    # @param epoch number of iteration to run
+    def run(self, epoch):
```

Comment:
I think this might be better named as iteration

Reply:
Done